### PR TITLE
wallet state machine, refactoring, tests

### PR DESCRIPTION
**Pull request by pragmaxim**
:wave: I was looking at https://github.com/ergoplatform/ergo/issues/1121 and I realized that for any kind of testing :
1. the `WalletActor`'s state must be extracted and passed through `context.become(walletInit(newState))`
2. this required a serious refactoring but I did my best to preserve all the contracts and business logic ... still the changes might seem radical as most of the methods cause a state transition `oldState -> newState` 

I know that this part is very sensitive + it had not many tests, so it might seem risky to release this. Maybe we could add more tests but as I'm new to Ergo, I'm not good at writing specs yet. I started covering it with tests but I will need much more time to learn the internal working...

### CODE REVIEW

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWallet.scala [hunk 1]
```diff
@@ -32,7 +32,7 @@ class ErgoWallet(historyReader: ErgoHistoryReader, settings: ErgoSettings)
   override type NVCT = this.type
 
   override val walletActor: ActorRef = {
-    val props = Props(classOf[ErgoWalletActor], settings, boxSelector, historyReader)
+    val props = Props(classOf[ErgoWalletActor], settings, new ErgoWalletServiceImpl, boxSelector, historyReader)
```
Comment (pragmaxim) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWallet.scala:35]: `ErgoWalletService` could be just `object` as it has no state, but this way we can mock it and test the logic of the Actor, there is still some business logic, for instance Recursive scanning that must be done by message passing to not block the actor ...

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                                         parameters: Parameters,
+                                         publicKeyAddresses: Seq[P2PKAddress]): Try[Seq[ErgoBoxCandidate]] =
+    Traverse[List].sequence {
+      requests.toList
+        .map {
+          case PaymentRequest(address, value, assets, registers) =>
+            Success(new ErgoBoxCandidate(value, address.script, fullHeight, assets.toColl, registers))
+          case AssetIssueRequest(addressOpt, valueOpt, amount, name, description, decimals, registers) =>
+            // Check that auxiliary registers do not try to rewrite registers R0...R6
+            if (registers.exists(_.exists(_._1.number < 7))) {
```
Comment (pragmaxim) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala:141]: I changed this from `if (registers.exists(_.forall(_._1.number < 7)))` as I assume we don't want any 0-7 register to be overriden. Is that correct or ?

File: src/it/resources/devnetTemplate.conf [hunk 3]
```diff
@@ -55,6 +55,6 @@ scorex {
   }
   restApi {
     bindAddress = "0.0.0.0:9051"
-    apiKeyHash = null
+    apiKeyHash = "324dcf027dd4a30a932c441f365a25e86b173defa4b8e58948253471b81b72cf"
```
Comment (pragmaxim) [inline src/it/resources/devnetTemplate.conf:58]: This is related to https://github.com/ergoplatform/ergo/issues/1316 ... when all the configs get merged, `apiKeyHash`  eventually turns into empty string which is sort of a bug ... Anyway integration tests of wallet require authentication with api key so I added it to be able to authenticate

**Review by kushti — CHANGES_REQUESTED** [review]
Please add ScalaDoc for new classes and their public methods.

Please do some functional testing. I am getting wallet being locked after unlock (via panel).

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala (hunk 2 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala:141]: Oh I guess you're right. Would be good to cover this with a test.

Comment (l15k4) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala:141]: I covered that already, I was just not sure if the AssetIssueRequest aux registers really cannot override any of the R0-R6 https://github.com/ergoplatform/ergo/pull/1311/files#diff-fca9ab9b22f435644bbc31e76ac9b4e4fb37e61555dd05e51a1c3763ad8f9ff1R90-R93

Comment (pragmaxim) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala:141]: I covered that already, I was just not sure if the AssetIssueRequest aux registers really cannot override any of the R0-R6 https://github.com/ergoplatform/ergo/pull/1311/files#diff-fca9ab9b22f435644bbc31e76ac9b4e4fb37e61555dd05e51a1c3763ad8f9ff1R90-R93

File: src/test/scala/org/ergoplatform/local/MempoolAuditorSpec.scala [hunk 4]
```diff
@@ -80,7 +80,7 @@ class MempoolAuditorSpec extends AnyFlatSpec with NodeViewTestOps with ErgoTestH
 
     getPoolSize shouldBe 1 // first tx removed from pool during node view update
 
-    scorex.core.utils.untilTimeout(cleanupDuration * 3, 100.millis) {
+    scorex.core.utils.untilTimeout(cleanupDuration * 4, 100.millis) {
```
Comment (pragmaxim) [inline src/test/scala/org/ergoplatform/local/MempoolAuditorSpec.scala:83]: @kushti You probably know but this test often fails due to busy Travis. It did not help though

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala (hunk 2 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSupport.scala:142]: Yes, that was an intention (aux registers shouldn't override R0-R6)

File: src/it/resources/devnetTemplate.conf (hunk 3 above)
Comment (kushti) [reply src/it/resources/devnetTemplate.conf:58]: Looks good

**Review by kushti — APPROVED** [review]

**Pull request merged**